File: app/src/main/java/com/liang/tind/leetcode/backtracking/LeetCode46.java
package com.liang.tind.leetcode.backtracking;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

/**
 * created by sherlock
 * <p>
 * date 2019/12/22
 * <p>
 * 给定一个没有重复数字的序列，返回其所有可能的全排列。
 * <p>
 * 示例:
 * <p>
 * 输入: [1,2,3]
 * 输出:
 * [
 * [1,2,3],
 * [1,3,2],
 * [2,1,3],
 * [2,3,1],
 * [3,1,2],
 * [3,2,1]
 * ]
 * <p>
 * 来源：力扣（LeetCode）
 * 链接：https://leetcode-cn.com/problems/permutations
 * 著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
 */
public class LeetCode46 {

    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        dfs(0, nums, new LinkedList<Integer>(), res);
        return res;
    }

    private void dfs(int cur, int[] nums, LinkedList<Integer> tracks, List<List<Integer>> res) {
        if (cur == nums.length) {
            res.add(new ArrayList<>(tracks));
        } else {
            for (int i = 0; i < nums.length; i++) {
                if (tracks.contains(nums[i])) continue;
                tracks.add(nums[i]);
                dfs(cur + 1, nums, tracks, res);
                tracks.removeLast();
            }
        }
    }

}

File: app/src/test/java/com/liang/tind/leetcode/backtrace/LeetCode46.kt
package com.liang.tind.leetcode.backtrace

import android.icu.lang.UCharacter.GraphemeClusterBreak.L
import org.junit.Assert
import org.junit.Test
import java.util.*
import kotlin.collections.ArrayList

/**
 * @author: lonnie.liang
 * date: 2022/05/20 23:33
 * https://leetcode.cn/problems/permutations/
 */
class LeetCode46 {

    @Test
    fun test() {

    }

    private val res = mutableListOf<List<Int>>()

    fun permute(nums: IntArray): List<List<Int>> {
        //给定一个不含重复数字的数组 nums ，返回其 所有可能的全排列 。你可以 按任意顺序 返回答案。
        //输入：nums = [1,2,3]
        //输出：[[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]

        //输入：nums = [0,1]
        //输出：[[0,1],[1,0]]

        //输入：nums = [1]
        //输出：[[1]]
        res.clear()
        backtracking(nums, LinkedList<Int>())
        return res
    }

    private fun backtracking(nums: IntArray, path: MutableList<Int>) {
        if (path.size == nums.size) {
            res.add(kotlin.collections.ArrayList<Int>(path))
            return
        }
        for (i in nums) {
            if (path.contains(i)) continue
            path.add(i)
            backtracking(nums, path)
            path.removeAt(path.lastIndex)
        }
    }
}